Feat: Implement Farms and Rewards Distribution

amusingaxl opened the pull request:
- feat: introduce the SDAO ERC-20 token
- test: fix SDAO token test suite
- chore: remove empty line from test for consistency
- fix: disable autorun for lint github action
- refactor: replace forge-std/Test with dss-test
- feat: implement rewards distribution mechanism

amusingaxl committed:
feat: implement rewards distribution mechanism

amusingaxl committed:
feat: add sanity checks to distribute method

hexonaut commented:
If this PR is dependant on the other one I would change the base from `master` to that other PR so there is not duplicate code in the diff.

The other PR is looking close to finished, so let's merge that before I review this.

amusingaxl committed:
fix: bad variable name

amusingaxl commented:
> If this PR is dependant on the other one I would change the base from `master` to that other PR so there is not duplicate code in the diff.
> 
> The other PR is looking close to finished, so let's merge that before I review this.

Oops, my bad... that's what I meant to do all along.
I was using Github CLI for this and it crashed unexpectedly, so when I executed it with the recovery file it must've not updated it.

amusingaxl commented:
This is now ready for review.

amusingaxl committed:
chore: add dss-vest dependency

amusingaxl committed:
tests: add RewardsDistribution initial test suite

amusingaxl committed:
chore: fix CI for multi Solidity versions

amusingaxl committed:
chore: fix-2 CI for multi Solidity versions

amusingaxl committed:
chore: fix-3 CI for multi Solidity versions

amusingaxl committed:
chore: fix-4 CI for multi Solidity versions

amusingaxl committed:
chore: fix-5 CI for multi Solidity versions

amusingaxl committed:
chore: fix-6 CI for multi Solidity versions

amusingaxl committed:
chore: fix-7 CI for multi Solidity versions

amusingaxl committed:
chore: fix-8 CI for multi Solidity versions

amusingaxl committed:
fix: put DeploySubProxy compatible into a multi-version repo structure

amusingaxl committed:
test: add file vestId sanity checks tests

amusingaxl committed:
refactor: reorganize interfaces and mocks for reusability

amusingaxl committed:
fix: keep track of the last distribution

`stakingRewards.lastUpdatedTime()` is not reliable because it is updated not only by `notifyRewardAmount`, but also by `stake` and `withdraw`.

The catch though is that the actual calculation depends on the underlying vesting stream, not the actual distribution. If the `vestId` changes, we need to reset the value of the last distribution time.

amusingaxl committed:
fix: Solidity version pragma notation

amusingaxl committed:
refactor: clean-up RewardsDistribution

amusingaxl committed:
refactor: reorganize `RewardsDistribution`

amusingaxl committed:
fix: use setCalc instead of direct assignment on constructor

amusingaxl committed:
fix: broken SDAOInvariants test suit

amusingaxl committed:
refactor: reorganize SDAO test suite

hexonaut commented:
Couple points before I start reviewing the code:

 * Why is `dss-vest` being imported? If you need it for tests you should either mock it for unit tests or use the deployed contract for integration tests.
 * I would split out the tests and scripts into `src`, `scripts`, `tests` in the root directory.

amusingaxl commented:
> Why is dss-vest being imported? If you need it for tests you should either mock it for unit tests or use the deployed contract for integration tests.

There's no deployed `DssVest` for `SDAO` tokens.

Not all integration tests need to be forking tests. That's more a convenience for complex setups, when it would be too hard to replicate the running system.

Mocking seems counter-productive for 2 reasons:

1. Deploying a `DssVest` is trivial.
2. `RewardsDistribution` is more of a glue component, so it makes little to no sens testing it in isolation. That's the reason why we're bringing the original `DssVest` in and will do the same once `StakingRewards` takes its final form to get rid of the mocks.

> I would split out the tests and scripts into src, scripts, tests in the root directory.

Is that a personal preference or something that impacts what's being delivered?

> Please try to avoid force-pushing PRs. If you want to merge in changes then do a merge instead of rebase.

That was due to a Github CLI fuck up. Since the actual review process hadn't started yet, no context was lost because of it. I'm aware of the consequences of force-push and made a tactical use of it.

amusingaxl committed:
docs: add build instructions

hexonaut commented on src/RewardsDistribution.sol:
```diff
@@ -0,0 +1,201 @@
+    function setVestId(uint256 _vestId) internal {
+        require(dssVest.valid(_vestId), "RewardsDistribution/invalid-vest-id");
+        require(dssVest.usr(_vestId) == address(stakingRewards), "RewardsDistribution/invalid-vest-usr");
+        require(dssVest.mgr(_vestId) == address(this), "RewardsDistribution/invalid-vest-mgr");
```
I don't think `mgr` does what you are assuming here. Without the vesting being restricted anyone can call `vest()` and thus bypass the constraints in this contract. This contract cannot call `vest()` if it is restricted. I think you need to modify how this works as having this distribution contract as the `usr()` of the vest and restricting it. The tokens can then be passed through to the rewards contract.

amusingaxl replied:
Gotcha... I believe I overlooked the actual implementation a bit too much.

This is actually an argument in favor of importing the real dependencies instead of mocking, because I'd probably build the mock with the wrong assumption.

Thanks for pointing this out.

hexonaut replied:
Fixed. Can't resolve comments.

hexonaut commented on src/DistributionCalc.sol:
```diff
@@ -0,0 +1,51 @@
+    }
+}
+
+contract ConstantDistribution is DistributionCalc {
+    function getAmount(
+        uint256 when,
+        uint256 prev,
+        uint256 tot,
+        uint256 fin,
+        uint256 clf
+    ) external pure returns (uint256) {
+        return (tot * (when - prev)) / (fin - clf);
+    }
+}
```
Vest is already a constant distribution and you can just call `vest(id)` without the amount to just forward the entire amount. For simplicity I would drop this calc, and in fact should probably just drop the external calc contracts altogether in favour of inheritance as it reduces the number of contracts to deploy which is always nice for maintenance. Default `RewardsDistribution` which can be inherited by `LinearRampUpRewardsDistribution` to override the amount to vest.

amusingaxl replied:
I understand this, however there's a case where it wouldn't work:

When `clf > bgn`, the total amount accrued up until the cliff would be disbursed right the way and from what I got from the requirements, this is something that we want to avoid.

Maybe for this specific use case we can assume `eta` will always be `0`, so yeah, this calc might be superfluous, but we should then require `clf == bgn` for it not to break.

However, I believe it would be a good idea to keep the flexibility for consistency and reusability.

---

Regarding changing the structure, I wouldn't mind, but I think there might an upside of preferring Composition to Inheritance here.

Originally, the stream rate for governance tokens would grow from 100k/year to 2MM, then it should remain constant as 4MM.

If we go with inheritance, we would need to replace the `RewardsDistribution` instance.

`distribute()` is probably going to be called by a keeper (be it a bot or in the Keep3r Network). This would mean that we'd have to add the new `RewardsDistribution` address and drop the old one.

If we keep the composition, the only thing we need to change is the `DistributionCalc`, so the `RewardsDistribution` address will remain the same, requiring no changes to the keepers.

hexonaut replied:
So we should not be over-engineering this to be the most flexible. The requirements are that we need two types of streams:

 * One that is constant the whole way through and does not use a cliff. So having that `require(clf == bgn)` as you say is the way I would go.
 * The other is one that starts at 0, ramps up to some number and stays constant from there on out. The `RewardDistribution` for that one should not have to be replaced and should automatically switch from linear increasing to constant distribution at the kink point.

Because this is all well-defined, we should use inheritance with a formula to determine the amount to send which matches the project's exact requirements. This is a case where you are tempted to make things ultra-flexible as an engineer (I do this too sometimes), but you need to try to avoid this instinct for SC dev. Mistakes are made when trying to go beyond what the exact requirements are. Simple is better.

amusingaxl replied:
> So we should not be over-engineering this to be the most flexible.

I beg to disagree this is over-engineering, it's a clear case of [a well known design pattern](https://en.wikipedia.org/wiki/Strategy_pattern), which is even [used across the MCD protocol](https://github.com/makerdao/dss/blob/master/src/abaci.sol).

I'm trying to optimize not only for development/deployment cost and complexity, but also for operational cost and complexity.

The solution I'm proposing may not be optimal from the former, but when combining it with the latter, I believe it reaches an optimal point, given the requirements presented so far.

In the end, it's one additional stateless read-only contract being deployed and provided as a constructor parameter. An operational error here would have the effect of distributing the wrong amount of rewards, but only until a new, correct contract is deployed. The risk seems pretty low.

> This is a case where you are tempted to make things ultra-flexible as an engineer (I do this too sometimes), but you need to try to avoid this instinct for SC dev. Mistakes are made when trying to go beyond what the exact requirements are. Simple is better.

We're in an environment of high uncertainty right now, and requirements change way too often. I am not deliberately trying to make things "ultra-flexible", but I believe some degree of flexibility is desired in the current context, otherwise we risk having to come back to the drawing board if requirements change, on a tighter schedule.

"Simple is better [than complex]", but also ["beautiful is better than ugly" and "explicit is better than implicit"](https://peps.python.org/pep-0020/).


> - The other is one that starts at 0, ramps up to some number and stays constant from there on out.

I thought the same initially, but from the last meeting we had, one of the rewards distribution streams would start at 100k/year and end at 2MM/year. After that, it should remain constant as 4MM/year. Please correct me if I'm wrong.

> The RewardDistribution for that one should not have to be replaced and should automatically switch from linear increasing to constant distribution at the kink point.

The way it's currently implemented, the vesting stream is king. The `RerwardsDistribution` contract ensures the full amount being vested ends up in the `StakingRewards` contract, no matter what. In other words, the vesting stream is the single source of truth. There's a clear way to implement the requirements:

1. Start with a vesting stream for the ramp up period and point the `RewardsDistribution` to it.
2. When we are ready to make the switch, create another vesting stream and update `RewardsDistribution` to account for the new stream and the new calc strategy.

The way you are imagining it, `Go from 100k to 2MM and then stay constant at 4MM after X happens` is a little bit harder to model with 1 contract + inheritance, since a `minRate` and a `maxRate` wouldn't suffice. We would have to have a third `rateAfterRampUp` variable.

Another issue is that we don't know exactly where in time the kink point should be, because it depends on a certain smart contract module being available, not a hard date AFAIK. This could be parametrizable, but that could bring even more operational complexity.

If we start to put more control into the `RewardsDistribution` contract itself, we'll need to ensure that either the vesting stream will comply with the requirements (i.e.: will it be able to distribute 4MM/year after kink point?), making the code more rigid and more complex.

The alternative would be to deploy a new base `RewardsDistribution`, but then we would have a different address, requiring any existing automation to update the reference to it.

In either case, there would be more moving parts.

hexonaut replied:
Yes you're right it looks like the requirements have changed slightly with 100k start. Nevertheless I would like to pursue an inheritance + put all the custom ramp-up logic inside `RewardsDistribution`. `dss-vest` streams should just be the amount once the ramp-up period ends. The abacus pattern I am against in both the way it was used in the clipper as well as this instance. Sometimes it can be useful, but not in these cases. If requirements change in the next few months we will just need to adjust the code, but probably only slightly.

hexonaut commented on src/RewardsDistribution.sol:
```diff
@@ -0,0 +1,201 @@
+        // If `lastVestedAt == 0`, it means it this is the first time we call `distribute` for the current `vestId`.
+        uint256 prev = lastVestedAt == 0 ? cliff : lastVestedAt;
+
+        uint256 amount = calc.getAmount(block.timestamp, prev, total, finish, cliff);
```
As I said above I think this can just be made into a function `_getVestAmount(...)` on a virtual function which returns `type(uint256).max` by default and can be overridden by `LinearRampUpDistribution` inheriting contract to define a linear ramp up with a timestamped kink so it can switch to a constant distribution after the ramp up period is done.

amusingaxl replied:
Replied above why I believe it would be beneficial to keep this in a separate contract.

hexonaut commented on src/RewardsDistribution.t.sol:
```diff
@@ -0,0 +1,253 @@
+    SDAO token;
+
+    uint256 totalRewards = 1 * 10 ** 18;
+    uint256 duration = 360 days;
```
Is this supposed to be a year? Would be 365 days then.

amusingaxl replied:
It doesn't make that much of a difference, but yeah, I'll change this.

hexonaut replied:
Fixed. Can't resolve comments.

hexonaut commented on src/RewardsDistribution.t.sol:
```diff
@@ -0,0 +1,253 @@
+        uint256 eta,
+        address mgr
+    ) internal returns (uint256 _vestId) {
+        return vest.create(usr, tot, bgn, tau, eta, mgr);
```
Should `vest.restrict(vestId)` after creating it.

amusingaxl replied:
Fixed.

hexonaut replied:
Fixed. Can't resolve comments.

hexonaut commented on src/RewardsDistribution.t.sol:
```diff
@@ -0,0 +1,253 @@
+        dist.distribute();
+
+        // Allow for 0,01% error tolerance due to rounding errors.
+        uint256 tolerance = 0.01e18;
```
This is a 1% error tolerance and not 1bps. I would reduce it to a smaller amount like 1bps as you say.

amusingaxl replied:
Fixed

hexonaut replied:
Fixed. Can't resolve comments.

hexonaut commented on src/RewardsDistribution.t.sol:
```diff
@@ -0,0 +1,253 @@
+        assertEq(token.balanceOf(address(farm)), 0);
+
+        vm.expectEmit(false, false, false, true, address(farm));
+        emit RewardAdded(totalRewards);
```
Maybe one of these event checks should be the `Distribute` event.

amusingaxl replied:
Yeah, `Distribute` makes more sense here.
Fixed.

hexonaut replied:
Fixed. Can't resolve comments.

hexonaut commented on src/RewardsDistribution.t.sol:
```diff
@@ -0,0 +1,253 @@
+        uint256 bgn = block.timestamp; // start immediately
+        uint256 tau = duration; // 1 year duration
+        uint256 eta = 0; // No cliff; start immediatebly
+        address mgr = address(dist);
```
`mgr` is just to `yank()` so this can be set to `address(0)`

amusingaxl replied:
Thanks for pointing that out, I somehow overlooked this and assumed something completely different.
It's fixed now.

hexonaut replied:
Fixed. Can't resolve comments.

hexonaut commented on src/RewardsDistribution.t.sol:
```diff
@@ -0,0 +1,253 @@
+        // `checkFileUint` increaments the current value of the param being modified. Since `vestId` is validated,
+        // we need to create a new one to make sure the `file`d param is valid.
+        _createVest();
+        checkFileUint(address(dist), "RewardsDistribution", ["vestId"]);
```
Would add a check here that `lastVestedAt` is set to 0 as well.

amusingaxl replied:
Fixed.

hexonaut replied:
Fixed. Can't resolve comments.

hexonaut commented on src/RewardsDistribution.t.sol:
```diff
@@ -0,0 +1,253 @@
+        checkFileAddress(address(dist), "RewardsDistribution", ["calc"]);
+    }
+
+    function testModifiers(address sender) private {
+        vm.assume(sender != address(this));
+
+        bytes4[] memory authedMethods = new bytes4[](1);
+        // authedMethods[0] = RewardsDistribution.mint.selector;
+
+        vm.startPrank(sender);
+        checkModifier(address(dist), "RewardsDistribution/not-authorized", authedMethods);
+    }
```
This can be removed as all `auth`s are already being checked.

amusingaxl replied:
Fixed.

hexonaut replied:
Fixed. Can't resolve comments.

hexonaut reviewed (changes requested):
General design looks good. There is one critical bug and some suggested simplifications for the calc. Once these are done I think it's good to merge this and pulling the whole thing together with a full deploy script and integration tests is the next step.

For deploy scripts I would recommend this pattern:

Seperate stand-alone libraries for deploy and init: https://github.com/makerdao/dss-direct-deposit/blob/master/src/deploy/D3MDeploy.sol / https://github.com/makerdao/dss-direct-deposit/blob/master/src/deploy/D3MInit.sol

Init should have no external dependencies other than interfaces. This makes it copy+paste into the spell.

There is then corresponding argument parsing wrappers: https://github.com/makerdao/dss-direct-deposit/blob/master/script/D3MDeploy.s.sol / https://github.com/makerdao/dss-direct-deposit/blob/master/script/D3MInit.s.sol

For the integration tests you can use the library logic and it will be the same for the spell so we don't get any surprises.

You can see an example of a spell using the library to init here: https://github.com/makerdao/spells-mainnet/tree/master/archive/2023-02-22-DssSpell

This saves a lot of time spell-crafting and reduces errors since there is no code changes from integration tests inside this repository.

amusingaxl committed:
refactor: make RewardsDistribution to work with restricted vesting streams

amusingaxl committed:
docs: add documentation regarding RewardsDistribution

hexonaut commented on src/interfaces/GemLike.sol:
```diff
@@ -0,0 +1,27 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity >=0.8.8;
+
+interface GemLike {
```
For this and the Vest interfaces you can use `dss-interfaces`: https://github.com/makerdao/dss-interfaces/blob/master/src/ERC/GemAbstract.sol and https://github.com/makerdao/dss-interfaces/blob/master/src/dss/VestAbstract.sol respectively.

I don't mind too much having the interfaces in here, but using `dss-interfaces` saves you some time. It's what it's there for. :)

amusingaxl replied:
Since it's only a couple of small interfaces, I rather have them inlined here than introduce a new dependency.

hexonaut replied:
`dss-interfaces` is actually already a dependancy of `dss-test` FYI, but I don't care much about this either way.

hexonaut commented on src/RewardsDistribution.t.sol:
```diff
@@ -29,225 +29,471 @@ import {DistributionCalc, LinearIncreasingDistribution, ConstantDistribution} fr
+    uint256 constant DEFAULT_TOTAL_REWARDS = 4_000_000 * 10 ** 18;
+    uint256 constant DEFAULT_DURATION = 365 days;
+    uint256 constant DEFAULT_CLIFF = 0;
+    uint256 constant DEFATUL_INITIAL_RATE = uint256(100_000 * 10 ** 18) / DEFAULT_DURATION;
```
Typo DEFATUL_INITIAL_RATE

amusingaxl replied:
Fixed.

hexonaut replied:
Can resolve.

hexonaut commented on src/DistributionCalc.sol:
```diff
@@ -16,29 +16,72 @@
+    uint256 public immutable initialRate;
+
+    /**
+     * @param _initialRate The initial raate to start the distribution.
```
Typo raate

amusingaxl replied:
Fixed

hexonaut replied:
Can resolve.

hexonaut commented on src/DistributionCalc.sol:
```diff
@@ -16,29 +16,72 @@
+ * @author amusingaxl
+ */
 contract LinearIncreasingDistribution is DistributionCalc {
+    /// @dev The initial raate to start the distribution.
```
Typo raate

amusingaxl replied:
Fixed

hexonaut replied:
Can resolve.

amusingaxl committed:
chore: fix typos

hexonaut commented:
Typo fixes look good. I'll continue to review after the change is made to inheritance with the ramp-up logic inside the `RewardsDistributor`.

amusingaxl committed:
fix: broken license header

amusingaxl committed:
refactor: rename `initialRate` to `startingRate`

amusingaxl committed:
refactor: simplify rewards distribution and distribution calc structure

amusingaxl committed:
fix: misplaced assignment of `lastVestedId`

amusingaxl committed:
fix: bad linear ramp up distribution formula

amusingaxl committed:
Feat: add farm

* chore: setup multi version & add farm

* choreL cleanup

* forge install: forge-std

* choreL cleanup

* chore: migrate to solc 0.6

* chore: migrate to solc 0.6

* forge install: openzeppelin-contracts

* chore: upgrade sol to 0.8

* feat: add referral code overload to stake method

* chore: remove vscode specific files.

Signed-off-by: Amusing Axl <112016538+amusingaxl@users.noreply.github.com>

* refactor: address issues pointed in the review

* chore: configure Github Actions to run on every pull request

* fix: bad Github Action config to run on every pull request

---------

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Amusing Axl <112016538+amusingaxl@users.noreply.github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Amusing Axl <112016538+amusingaxl@users.noreply.github.com>

amusingaxl committed:
refactor: use real StakingRewards in tests

amusingaxl committed:
refacotr: reorganize repo structure

amusingaxl committed:
chore: fix license for Solidity files

amusingaxl committed:
refactor: inline RewardsDistributionRecipient into StakingRewards

amusingaxl committed:
refactor: add missing event to setRewardsDistribution

amusingaxl committed:
refactor: make additional improvements over the original Synthetix code

amusingaxl committed:
chore: add file permissions config to foundry.toml

hexonaut commented on script/DeploySubProxy.s.sol:
```diff
@@ -16,7 +16,7 @@
-import {SubProxy} from "../SubProxy.sol";
+import {SubProxy} from "../src/SubProxy.sol";
 
 contract DeploySubProxy is Script {
```
Missing `vm.stopBroadcast()` btw.

amusingaxl replied:
This file is going to be replaced in another PR which is currently in draft.

hexonaut replied:
Cool feel free to resolve. Btw would be nice for you to give me resolve permissions. :)

hexonaut commented on src/VestedRewardsDistribution.sol:
```diff
@@ -140,12 +143,14 @@ contract RewardsDistribution {
+        require(dssVest.clf(_vestId) == dssVest.bgn(_vestId), "VestedRewardsDistribution/invalid-vest-cliff");
 
         if (dssVest.res(_vestId) == 0) {
             dssVest.restrict(_vestId);
```
I think it's better practise to use a require for this restrict being set to true instead of silently fixing it. The spell should be designed such that this errors if the person is not configuring the vest correctly in the first place.

amusingaxl replied:
Good point.

amusingaxl replied:
Fixed.

hexonaut replied:
This is fixed. You can close - I don't have permissions.

hexonaut commented on src/VestedRewardsDistribution.sol:
```diff
@@ -162,48 +167,65 @@ contract RewardsDistribution {
+        // We ensure that `clf == bgn`, so we could use either one.
+        uint256 clf = dssVest.clf(vestId);
+        // If `lastVestedAt == 0`, it means it this is the first time we call `distribute` for the current `vestId`.
+        uint256 prev = lastVestedAt == 0 ? clf : lastVestedAt;
```
Should add a require that `block.timestamp >= prev`.

amusingaxl replied:
Yeah, that was safeguarded before within the `DistributionCalc` logic, but now there might be a problem.

Actually, I think the rule should probably be `block.timestamp > prev`.

amusingaxl replied:
Errm, actually I think this is not needed.
If you call `distribute()` once, `lastVestedAt` is updated to the current block timestamp.

If you call it again within the same block, it will fail at:
```
require(amount > 0, "VestedRewardsDistribution/no-pending-amount");
```
on L194.

However, we might want to add the check for clarity.

amusingaxl replied:
FYI I added the check.

hexonaut replied:
This can still set `prev` to be a value greater than `block.timestamp` if we are before the `clf` time. It's better to explicitly fail in this case than have subtraction underflow.

amusingaxl replied:
Fixed.

hexonaut commented on src/synthetix/StakingRewards.sol:
```diff
@@ -0,0 +1,197 @@
+import {IStakingRewards} from "./interfaces/IStakingRewards.sol";
+import {Pausable} from "./utils/Pausable.sol";
+
+// https://docs.synthetix.io/contracts/source/contracts/stakingrewards
```
Link doesn't work.

amusingaxl replied:
See below.

hexonaut replied:
Ok that's fine. I see the note at the top. Feel free to resolve this.

hexonaut commented on src/synthetix/interfaces/IStakingRewards.sol:
```diff
@@ -0,0 +1,58 @@
+
+import {IERC20} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol";
+
+// https://docs.synthetix.io/contracts/source/interfaces/istakingrewards
```
Link doesn't work.

amusingaxl replied:
This link is broken in the original code as well. We decided not change what is not required to make the code compile to minimize the diff.

hexonaut replied:
Ok that's fine. I see the note at the top. Feel free to resolve this.

hexonaut commented on src/DistributionCalc.sol:
```diff
@@ -43,52 +42,34 @@ interface DistributionCalc {
-        uint256 clf
-    ) external pure returns (uint256) {
-        return (tot * (when - prev)) / (fin - clf);
+            ((tot - startingRate * duration) *
+                ((when - clf) ** 2 - (prev - clf) ** 2) +
+                (startingRate * interval * divisor)) / (divisor);
```
A derivation of how this is calculated would be nice.

amusingaxl replied:
Added the comment with a summarized derivation of the formula.
We can reference the technical doc in this repo after it's fully reviewed to give readers the full context.

hexonaut replied:
Ok I see it's been added. Feel free to resolve this.

amusingaxl committed:
docs: add reference to the original `StakingRewards` contract

amusingaxl committed:
docs: add `LinearRampUp` formula derivation to natspec.

hexonaut reviewed (commented):
Mostly looks good. I still need to review the linear ramp-up is correct.

Can be added in another PR, but would like to see full e2e integration tests.

amusingaxl committed:
refactor: stop silently fixing unrestricted vesting streams

Add a sanity check in `_setVestId` to prevent unrestricted streams from being used.

amusingaxl committed:
refactor: rename `lastVestedAt` to `lastDistributedAt`

The new name conveys the meaning better.

amusingaxl commented:
> Can be added in another PR, but would like to see full e2e integration tests.

What are the components that you want involved in these tests?
`VestedRewardsDistribution.t.sol` can be thought of as an e2e test suite, since it uses the real `DssVest`, `VestedRewardsDistribution` and `StakingRewards` contracts.

I think the only missing step would be testing if the staking users get the appropriate amounts, but that is pretty much a given, since we are reusing a glorified contract.

hexonaut commented:
Yes the VestedRewardsDistribution is pretty close, but usually I like to see absolutely everything including setup in the pause proxy, etc.

hexonaut commented on src/DistributionCalc.sol:
```diff
@@ -0,0 +1,134 @@
+     *
+     *     r(prev, when)  =  tot - s(fin  -  clf)                2                   2
+     *                       -------------------- [(when  -  clf)   -  (prev  -  clf)  ]  +  s(when  -  prev)
+     *                          2 (fin  -  clf)
```
I think this `2` in the denominator should of canceled out with the `2` in the numerator for `k`. This is missing the square on the `(fin - clf)` term in the denominator. Maybe the `2` was meant to be a square instead of a multiplication?

amusingaxl replied:
Yeah, it looks like a bug in the LaTeX to ASCII tool I was using. Good catch. Will fix it.

amusingaxl replied:
Fixed

hexonaut replied:
LGTM, can resolve

amusingaxl committed:
docs: fix equation in natspec

There was a typo in one of the equations in the description.
The formula implemented in the code was correct.

hexonaut commented on src/DistributionCalc.sol:
```diff
@@ -0,0 +1,135 @@
+        uint256 divisor = duration ** 2;
+
+        return
+            ((tot - startingRate * duration) *
```
`tot >= startingRate * duration` as long as the vest is configured properly. Maybe it would be good to include a formula in the readme or something how to calculate what each of the vest parameters should be based on the desired rate ramp-up. IE add a formula for calculating `tot = ...`, etc so there are no mistakes configuring the vest.

amusingaxl replied:
The formula and how we got to it is in the technical doc, but I'll make sure to add the final formula to the README.

amusingaxl replied:
Done. Please check the updated README.
I also changed the integration test in `VestedRewardsDistribution.t.sol` to calculate the total amount to be distributed the same way.

hexonaut commented on src/VestedRewardsDistribution.t.sol:
```diff
@@ -0,0 +1,545 @@
+        }
+
+        // Check if balance grows faster every time:
+        for (uint256 i = 1; i < deltas.length; i++) {
+            assertGt(
+                deltas[i],
+                deltas[i - 1],
+                string.concat("Bad balance change between #", toString(i - 1), " and #", toString(i))
+            );
+        }
```
Can we not only check that this is increasing, but also that it is increasingly roughly linearly?

amusingaxl replied:
Done.

amusingaxl committed:
test: improve accuracy of linear ramp-up test

Previously we were assessing only if the distribution amount would increase over time. We changed the test to check if the increase is actually linear.

amusingaxl committed:
docs: add more component descriptions to README

amusingaxl committed:
test: add check on undistributed amounts

amusingaxl committed:
refactor: improve distribution amount logic

amusingaxl committed:
refactor: add explicit error when `LinearRampUp` starting rate is set too high

amusingaxl committed:
docs: improve content on LinearRampUp natspec

hexonaut commented on src/VestedRewardsDistribution.sol:
```diff
@@ -181,45 +181,51 @@ contract VestedRewardsDistribution {
-            return type(uint256).max;
+    function _getAmount() internal view returns (uint256) {
+        // We ensured that `clf == bgn`, so we could use either one.
+        uint256 clf = dssVest.clf(vestId);
+        // If the current timestamp is before the cliff, there is nothing to be distriubted.
+        if (block.timestamp < clf) {
+            return 0;
+        }
```
This check is redundant with the unpaid check below. You can just remove it.

amusingaxl replied:
Yeah, good point.

amusingaxl replied:
It's fixed now.

hexonaut reviewed (changes requested):
One point, but rest looks good.

amusingaxl committed:
docs: add charts to LinearRampUp natspec

Added missing charts to the doc comment to make it easier to understand the formulas

amusingaxl committed:
refactor: remove redundant check from `_getAmount()`

amusingaxl committed:
docs: fix wrong variable name in comments

hexonaut reviewed (approved):
LGTM

0xdecr1pto reviewed (approved):
LGTM
Nice job 🥇
